api/session.ts

```diff
@@ -0,0 +1,429 @@
+		}
+		const timeout = (curTime + SID_DURATION);
+		const ip = this.dispatcher.getIp();
+		const sidhash = await this.getSid();
+		const res = await sessions.insert({userid, sid: sidhash, time: time(), timeout, ip});
+		this.sidhash = sidhash;
+		this.session = res.insertId || 0;
+		this.dispatcher.user.login(name);
+		this.updateCookie();
+		return this.dispatcher.user;
+	}
+	async logout() {
+		if (!this.session) return false;
+		await sessions.delete(this.session);
+		this.sidhash = '';
+		this.session = 0;
+		this.deleteCookie();
+		this.dispatcher.user.logout();
+	}
+	updateCookie() {
+		const name = this.getName();
+		if (toID(name) === 'guest') return;
+		if (!this.sidhash) {
+			return this.deleteCookie();
+		}
+		const rawsid = encodeURIComponent([name, this.session, this.sidhash].join(','));
+		this.dispatcher.setHeader(
+			'Set-Cookie',
+			`sid=${rawsid}; Max-Age=31363200; Domain=${Config.routes.root}; Path=/; Secure; SameSite=None`
+		);
+	}
+	async getAssertion(
+		userid: string,
+		challengekeyid: number = -1,
+		user: User | null,
+		challenge: string = '',
+		challengeprefix: string = ''
+	) {
+		if (userid.startsWith('guest')) {
+			return ';;Your username cannot start with \'guest\'.';
+		} else if (userid.length > 18) {
+			return ';;Your username must be less than 19 characters long.';
+		} else if (!Session.isUseridAllowed(userid)) {
+			return ';;Your username contains disallowed text.';
+		}
+		let data = '';
+		const ip = this.dispatcher.getIp();
+		let forceUsertype: string | false = false;
+		if (!user) user = this.dispatcher.user;
+		if (ip.includes(Config.autolockip) || Config.autolockip?.includes(ip)) {
+			forceUsertype = '5';
+		}
+		let userType = '';
+		const userData = await user.getData();
+		const {banstate, registertime: regtime} = userData;
+		const server = this.dispatcher.getServer();
+		const serverHost = server?.server || "sim3.psim.us";
+
+		if (user.id === userid && user.loggedin) {
+			// already logged in
+			userType = '2';
+			if (Config.sysops?.includes(user.id)) {
+				userType = '3';
+			} else {
+				const customType = Config.getUserType?.call(this, user, banstate, server);
+				if (forceUsertype) {
+					userType = forceUsertype;
+				} else if (customType) {
+					userType = customType;
+				} else if (banstate <= -10) {
+					userType = '4';
+				} else if (banstate >= 100) {
+					return ';;Your username is no longer available.';
+				} else if (banstate >= 40) {
+					if (server?.server === 'sim3.psim.us') {
+						userType = '40';
+					} else {
+						userType = '2';
+					}
+				} else if (banstate >= 30) {
+					userType = '6';
+				} else if (banstate >= 20) {
+					userType = '5';
+				} else if (banstate === 0) {
+					if (regtime && time() - regtime > (7 * 24 * 60 * 60)) {
+						const ladders = await ladder.selectOne('formatid', 'userid = ? AND w != 0', [userid]);
+						if (ladders) {
+							userType = '4';
+							void users.update(userid, {banstate: -10});
+						}
+					}
```

Comment:
```suggestion
					if (regtime && time() - regtime > (7 * 24 * 60 * 60)) {
						// update autoconfirmed status if necessary
						const ladders = await ladder.selectOne('formatid', 'userid = ? AND w != 0', [userid]);
						if (ladders) {
							userType = '4';
							void users.update(userid, {banstate: -10});
						}
					}
```
The magic numbers for user types (which tbh should probably be top-level `const`s or attributes of a top-level const object) aren't documented, so I think a comment for this would be good.